packages/sdk/src/streamStateView_Space.ts

```diff
@@ -150,6 +131,62 @@ export class StreamStateView_Space extends StreamStateView_AbstractContent {
         }
     }
 
+    private decryptionInProgress: Promise<ChunkedMedia | undefined> | undefined
+    private decryptionQueue: Promise<void> = Promise.resolve()
+    private latestEncryptedImage: EncryptedData | undefined
+
+    public async getSpaceImage(): Promise<ChunkedMedia | undefined> {
+        // Take a snapshot of the current encrypted image
+        const currentEncryptedImage = this.encryptedSpaceImage
+
+        // If the current image is already being decrypted, return the ongoing decryption promise
+        if (this.decryptionInProgress && currentEncryptedImage === this.latestEncryptedImage) {
+            return this.decryptionInProgress
+        }
+
+        if (currentEncryptedImage) {
+            const spaceAddress = contractAddressFromSpaceId(this.streamId)
+            const context = spaceAddress.toLowerCase()
+
+            // Update the latest encrypted image
+            this.latestEncryptedImage = currentEncryptedImage
+
+            // Chain the decryption operations sequentially
+            this.decryptionQueue = this.decryptionQueue.then(async () => {
+                // Assign the decryption process to decryptionInProgress and return void to avoid ESLint warnings
+                this.decryptionInProgress = this.performDecryption(context, currentEncryptedImage)
+                return this.decryptionInProgress.then(() => undefined) // Ensure void return
+            })
+
+            await this.decryptionQueue
+
+            // Always return the last known good state of spaceImage, even if it wasn't updated
+            return this.spaceImage
+        }
+
+        return this.spaceImage
+    }
+
+    private async performDecryption(
+        context: string,
+        currentEncryptedImage: EncryptedData,
+    ): Promise<ChunkedMedia | undefined> {
+        try {
+            const plaintext = await decryptDerivedAESGCM(context, currentEncryptedImage)
+
+            // Ensure the state is still relevant before updating
+            if (this.encryptedSpaceImage === currentEncryptedImage) {
+                this.spaceImage = ChunkedMedia.fromBinary(plaintext)
+                this.encryptedSpaceImage = undefined
+            }
+
+            return this.spaceImage
+        } finally {
+            // Clear the in-progress promise once it resolves or rejects
+            this.decryptionInProgress = undefined
+        }
+    }
+
```

Comment:
I'm not sure chat gpt is the best at this. Here what about this? https://github.com/river-build/river/pull/629/files

Reply:
first caller might get spaceImage undefined if it completed the decryption, but a call to applySnapshot had updated the value of encryptedSpaceImage during its await.

Reply:
the decryptionQueue ensures that decryption requests are handled sequentially. It prevents overlapping decryptions. More orderly execution at runtime. The caller should get a predictable result; not undefined.

Reply:
> irst caller might get spaceImage undefined if it completed the decryption, but a call to applySnapshot had updated the value of encryptedSpaceImage during its await.
Pretty sure that's not the case. Can you look again.

Reply:
> the decryptionQueue ensures that decryption requests are handled sequentially. It prevents overlapping decryptions. More orderly execution at runtime. The caller should get a predictable result; not undefined.

1) I don't think that's a problem
2) i don't think your queue is doing what you think it is. The way I read it, it's also doing overlapping decryptions.

Reply:
your code will return an undefined space image for the first caller if the source is updated between calls.